flow/core/kernel/scenario/aimsun.py

```diff
@@ -73,39 +78,80 @@ def generate_network(self, scenario):
+                f.write(template_path)
 
         # start the aimsun process
         aimsun_call = [aimsun_path, "-script", script_path]
         self.aimsun_proc = subprocess.Popen(aimsun_call)
 
         # merge types into edges
         if scenario.net_params.osm_path is None:
-            for i in range(len(scenario.edges)):
-                if 'type' in scenario.edges[i]:
-                    for typ in scenario.types:
-                        if typ['id'] == scenario.edges[i]['type']:
-                            new_dict = deepcopy(typ)
-                            new_dict.pop("id")
-                            scenario.edges[i].update(new_dict)
-                            break
-
-            self._edges = {}
-            for edge in deepcopy(scenario.edges):
-                edge_name = edge['id']
-                self._edges[edge_name] = {}
-                del edge['id']
-                self._edges[edge_name] = edge
+            if scenario.net_params.template is None:
+                for i in range(len(scenario.edges)):
+                    if 'type' in scenario.edges[i]:
+                        for typ in scenario.types:
+                            if typ['id'] == scenario.edges[i]['type']:
+                                new_dict = deepcopy(typ)
+                                new_dict.pop("id")
+                                scenario.edges[i].update(new_dict)
+                                break
+
+                self._edges = {}
+                for edge in deepcopy(scenario.edges):
+                    edge_name = edge['id']
+                    self._edges[edge_name] = {}
+                    del edge['id']
+                    self._edges[edge_name] = edge
+
+                # list of edges and internal links (junctions)
+                self._edge_list = [
+                    edge_id for edge_id in self._edges.keys()
+                    if edge_id[0] != ':'
+                ]
+                self._junction_list = list(
+                    set(self._edges.keys()) - set(self._edge_list))
+
+            else:
+                # load scenario from template
+                scenario_file = "flow/core/kernel/scenario/scenario_data.json"
+                filepath = os.path.join(config.PROJECT_PATH, scenario_file)
+
+                while not os.path.exists(filepath):
+                    time.sleep(0.5)
+
+                # file exists, wait to make sure all the data has been written
+                # TODO create a blank file when load.py is done writing
+                # in case 3 seconds isn't enough
+                time.sleep(3.0)
+
+                with open(filepath) as f:
+                    time.sleep(0.5)
+                    content = json.load(f)
+
+                self._edges = content['sections']
+                self._edge_list = self._edges.keys()
+                self._junction_list = content['turnings']
+                # TODO load everything that is in content into the scenario
```

Comment:
what still needs to be done for this TODO? Can we list them (will help out for later)

Reply:
We have the following information we can make use of (and we could possibly get more depending on needs):

- centroids (input or output)
- connections (junctions between centroids and sections)
- nodes (ie intersection, list of turnings)
- sections (length, max speed, number of lanes)
- turnings (origin and destination sections, origin and destination lanes, length, max speed)
(+ the names and IDs of everything)

The length of sections and turnings are approximations; the only thing I found to make it more precise would be to manually compute it from the raw polygon. 

We already have the edges (sections). We would need to load the centroids and connections, that could work as inflow/outflow somehow (although there are no real inflows in Aimsun). 

Then there are the junctions, defined in the rest of the code as `self._junction_list = list(set(self._edges.keys()) - set(self._edge_list))`, so just a list of section IDs, while what we retrieve is a list of origin lane/destination lane pairs -> we probably need to adapt the code so it can handle the two types of junctions, or directly upgrade junctions from a list of IDs to a list of pair of lanes (or change the structure even deeper and have for each lane/section a list of the possible next lane/section destinations).